src/python/grpcio/grpc/_cython/_cygrpc/aio/completion_queue.pyx.pxi

```diff
@@ -0,0 +1,97 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+cdef gpr_timespec _GPR_INF_FUTURE = gpr_inf_future(GPR_CLOCK_REALTIME)
+
+
+def _handle_callback_wrapper(CallbackWrapper callback_wrapper, int success):
+    CallbackWrapper.functor_run(callback_wrapper.c_functor(), success)
+
+
+cdef class BaseCompletionQueue:
+
+    async def shutdown(self):
+        raise NotImplementedError()
+
+    cdef grpc_completion_queue* c_ptr(self):
+        return self._cq
+
+
+cdef class PollerCompletionQueue(BaseCompletionQueue):
+
+    def __cinit__(self):
+        self._cq = grpc_completion_queue_create_for_next(NULL)
+        self._shutdown = False
+        self._shutdown_completed = asyncio.get_event_loop().create_future()
+        self._poller_thread = threading.Thread(target=self._poll_wrapper, daemon=True)
+        self._poller_thread.start()
+
+    cdef void _poll(self) except *:
+        cdef grpc_event event
+        cdef CallbackContext *context
+        cdef object waiter
+
+        while not self._shutdown:
+            with nogil:
+                event = grpc_completion_queue_next(self._cq,
+                                                _GPR_INF_FUTURE,
+                                                NULL)
+
+            if event.type == GRPC_QUEUE_TIMEOUT:
+                raise AssertionError("Core should not return timeout error!")
+            elif event.type == GRPC_QUEUE_SHUTDOWN:
+                self._shutdown = True
+                aio_loop_call_soon_threadsafe(self._shutdown_completed.set_result, None)
+            else:
+                context = <CallbackContext *>event.tag
+                aio_loop_call_soon_threadsafe(
+                    _handle_callback_wrapper,
+                    <CallbackWrapper>context.callback_wrapper,
+                    event.success)
+
+    def _poll_wrapper(self):
+        self._poll()
+
+    async def shutdown(self):
+        grpc_completion_queue_shutdown(self._cq)
+        await self._shutdown_completed
+        grpc_completion_queue_destroy(self._cq)
+        self._poller_thread.join()
+
+
+cdef class CallbackCompletionQueue(BaseCompletionQueue):
+
+    def __cinit__(self):
+        self._shutdown_completed = grpc_aio_loop().create_future()
+        self._wrapper = CallbackWrapper(
+            self._shutdown_completed,
+            CQ_SHUTDOWN_FAILURE_HANDLER)
+        self._cq = grpc_completion_queue_create_for_callback(
+            self._wrapper.c_functor(),
+            NULL
+        )
+
+    async def shutdown(self):
+        grpc_completion_queue_shutdown(self._cq)
+        await self._shutdown_completed
+        grpc_completion_queue_destroy(self._cq)
+
+
+cdef BaseCompletionQueue create_completion_queue():
+    if grpc_aio_engine is AsyncIOEngine.CUSTOM_IO_MANAGER:
+        return CallbackCompletionQueue()
+    elif grpc_aio_engine is AsyncIOEngine.POLLER:
+        return PollerCompletionQueue()
```

Comment:
Now I can see that we are creating one cq per channel, server and then one thread for attending asynchronously the events.

Would it make sense of having only 1 CQ and 1 thread? This would reduce the resources needed and due to the nature of Python and Asyncio might perform even better. IIRC the node implementation uses only one CQ for all of the channels/servers.

Reply:
Good idea! I missed this important feature.

Reply:
I created https://github.com/grpc/grpc/issues/22288 to track this feature. Can we implement it in a separate PR? It changes our way of plumbing completion queue objects to each channel and each server. It seems cleaner to split into two PR.

Reply:
yes ofc.